Add config flow to the Onkyo component and change to local push

klaashoekstra94 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Rename service `media_player.onkyo_select_hdmi_output` to `onkyo.select_hdmi_output`

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This is a follow-up of the closed PR [#48640](https://github.com/home-assistant/core/pull/48640).
This PR adds a config flow for the Onkyo component. The component
now uses the pyeiscp library to make it local push instead of polling.
This PR also includes setting the sound mode on supported devices.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: [#20251](https://github.com/home-assistant/home-assistant.io/pull/20251)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository] [#20251](https://github.com/home-assistant/home-assistant.io/pull/20251)

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [x] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

klaashoekstra94 committed:
Add config flow to the Onkyo component

klaashoekstra94 commented:
I just thought of some more changes to increase the quality scale for the integration, which I'll soon add to this PR. But it would be great if someone could already start reviewing this.

frenck commented:
> But it would be great if someone could already start reviewing this.

At this point the builds are failing, please make sure the builds pass.

klaashoekstra94 committed:
Fixed Pylint issues

frenck commented on homeassistant/components/onkyo/__init__.py:
```diff
@@ -1 +1,71 @@
+
+    undo_listener = entry.add_update_listener(update_listener)
+
+    hass.data[DOMAIN][entry.entry_id] = {
```
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
```

frenck commented on homeassistant/components/onkyo/__init__.py:
```diff
@@ -1 +1,71 @@
+UNDO_UPDATE_LISTENER: str = "undo_update_listener"
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the Onkyo component."""
+    # Initiate the onkyo name space as a dict.
+    hass.data.setdefault(DOMAIN, {})
+
+    return True
+
+
```
```suggestion
```

frenck commented on homeassistant/components/onkyo/__init__.py:
```diff
@@ -1 +1,71 @@
+        UNDO_UPDATE_LISTENER: undo_listener,
+    }
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
```
```suggestion
    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```

frenck commented on homeassistant/components/onkyo/__init__.py:
```diff
@@ -1 +1,71 @@
+    if not receiver.online and not await receiver.async_connect():
+        raise ConfigEntryNotReady(f"Failed to connect to {receiver.name}")
+
+    undo_listener = entry.add_update_listener(update_listener)
```
```suggestion
    entry.async_on_unload(entry.add_update_listener(async_reload_entry))
```

frenck commented on homeassistant/components/onkyo/__init__.py:
```diff
@@ -1 +1,71 @@
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        CONF_RECEIVER: receiver,
+        UNDO_UPDATE_LISTENER: undo_listener,
```
```suggestion
```

frenck commented on homeassistant/components/onkyo/__init__.py:
```diff
@@ -1 +1,71 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
```
```suggestion
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```

frenck commented on homeassistant/components/onkyo/__init__.py:
```diff
@@ -1 +1,71 @@
+        )
+    )
+
+    entry_data = hass.data[DOMAIN][entry.entry_id]
+    entry_data[UNDO_UPDATE_LISTENER]()
+    entry_data[CONF_RECEIVER].disconnect()
```
```suggestion
    hass.data[DOMAIN][entry.entry_id][CONF_RECEIVER].disconnect()
```

frenck commented on homeassistant/components/onkyo/config_flow.py:
```diff
@@ -0,0 +1,235 @@
+    DEFAULT_SOURCES,
+    DISCOVER_TIMEOUT,
+)
+from .const import DOMAIN  # pylint:disable=unused-import
```
```suggestion
from .const import DOMAIN
```

frenck commented on homeassistant/components/onkyo/config_flow.py:
```diff
@@ -0,0 +1,235 @@
+
+        settings_schema = vol.Schema(
+            {
+                vol.Optional(
+                    CONF_MAX_VOLUME,
+                    default=self.config_entry.options.get(
+                        CONF_MAX_VOLUME, DEFAULT_MAX_VOLUME
+                    ),
+                ): vol.All(vol.Coerce(int), vol.Range(min=1)),
+                vol.Optional(
+                    CONF_ENABLED_SOURCES,
+                    default=list(
+                        self.config_entry.options.get(
+                            CONF_SOURCES, DEFAULT_SOURCES
+                        ).keys()
+                    ),
+                ): cv.multi_select(
+                    {
+                        key: f"{source_name} ({key})"
+                        for key, source_name in sources.items()
+                    }
+                ),
```
These are device-level options? Not integration options?

klaashoekstra94 replied:
I didn't know there's a difference, but these are indeed device-level options and may be different per added device. Do I have to change anything?

frenck replied:
I think max volume and such are device setting that can be just entities?

Enabled sources, I'm not sure, do we need it even? Why not enable all by default?

klaashoekstra94 replied:
What do you mean can be just entities? Users can use max_volume to limit the max volume of the receiver if 100% is too loud. So I think they should be able to configure it as an option. I think @OnFreund can agree with me here, he also helped out on the first [PR](https://github.com/home-assistant/core/pull/33694)

In the current integration the sources can also be set so I don't want to drop the functionality. It's also more clear to only see the inputs of the receiver you use.

frenck replied:
> What do you mean can be just entities? Users can use max_volume to limit the max volume of the receiver if 100% is too loud.

If the receiver has a max/min volume setting, it can be represented as an entity (number entity). 

> In the current integration the sources can also be set so I don't want to drop the functionality.

How would that drop functionality? The sources are still available.

klaashoekstra94 replied:
I've been playing around with changing the maximum volume to a number entity, but I'm not sure about a couple of things. The maximum volume of a receiver is either 80 or 200, depending on the model, but there is no easy way to read that info from the receiver, so the user should set it per device. 

But if you never want to have your receiver at max volume, you can also set the maximum volume lower so that HA will not set the volume above the set level.

Should all receivers then have a maximum volume number entity that ranges from 0 to 200? But it would not make sense to make this value higher then 80 on a receiver which max volume is 80. But if we want the range to stop at the max volume of the receiver, we're gonna need an option in the device option flow to set the max of the max volume range. 

Maybe @OnFreund has an idea which works best, keeping the max volume an option as it is already, or changing it to a number entity that always ranges from 0 to 200, or also make an option to set the max range of the number entity

frenck replied:
> we're gonna need an option in the device option flow to set the max of the max volume range.

Home Assistant doesn't have a concept of device option flows.

klaashoekstra94 replied:
Sorry, I should use the correct naming, I mean an integration configuration option, just like the max volume is implemented now.

frenck replied:
>  integration configuration option

Option flows for configurations should not really be used for "device" configuration.

klaashoekstra94 replied:
Okay thank you, shall I then just create a number entity for each device, although it will probably be set once?

frenck commented on homeassistant/components/onkyo/config_flow.py:
```diff
@@ -0,0 +1,235 @@
+from .const import DOMAIN  # pylint:disable=unused-import
+
+
+class OnkyoConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
Import of the existing YAML configuration is missing?
We should migrate existing users of the integration.

klaashoekstra94 replied:
I'm working on this now, am I correct that I should implement the `async_step_import` in the config flow? And should this directly create entries with the options from the old config?

frenck replied:
Correct :)

klaashoekstra94 replied:
Next question, how do I create multiple entries from one flow? All examples I see return self.async_create_entry, which only creates a single entry.

frenck replied:
> Next question, how do I create multiple entries from one flow?

You don't. Instead, call start the import flow multiple times.

klaashoekstra94 replied:
How do I start the import flow, isn't this done automatically when there is an entry in the yaml file?
In the current onkyo integration, all configuration variables are optional, so this is a valid yaml entry:
```
media_player:
  - platform: onkyo
```
which will discover and add all available receivers. How do I start multiple import flows from this one line?

frenck replied:
Here is an example on how to import an entry.

https://github.com/home-assistant/core/blob/606989917954b6549efa35f56a589f07357c57e8/homeassistant/components/stookalert/binary_sensor.py#L36-L57

`hass.config_entries.flow.async_init` can be called multiple times if needed.

klaashoekstra94 replied:
Thanks a lot for the help! So I should add this specific part to media_player.py?

frenck commented on homeassistant/components/onkyo/config_flow.py:
```diff
@@ -0,0 +1,235 @@
+            )
+            return await self.async_step_connect()
+
+        select_scheme = vol.Schema(
+            {
+                vol.Required("select_receiver"): vol.In(
+                    [conn.name for conn in self._discovered_connections]
+                )
+            }
+        )
```
So there is no option to add receivers manually... so why not have the integration add all discovered receivers to Home Assistant automatically?

That way, we don't have the ask the user anything? (See for example, Sonos & Cast integrations that do this).

klaashoekstra94 replied:
I'm now also working on an option to manually add a host, but it requires some changes in the library used.

frenck replied:
Should we mark this PR draft until that is done?

frenck reviewed (commented):
Some initial comments, mainly limited to init & config flow at first.

I think the biggest part missing at this point, is the migration path for existing users.

klaashoekstra94 committed:
Refactor __init__.py, get 100% cov on config_flow

klaashoekstra94 committed:
Improve disconnect handling

klaashoekstra94 committed:
Add manual config flow

klaashoekstra94 committed:
Add import flow step to migrate old configs

klaashoekstra94 committed:
Improve connecting to a receiver

klaashoekstra94 committed:
Update pyeiscp version

klaashoekstra94 committed:
Change max volume option to number entity

klaashoekstra94 commented:
I changed the max volume option from an integration option to a number entity per device. I use a `RestoreEntity` to make sure the value persists, since there is no way to get the information from the receiver.

Also, I think the breaking-change label can be removed, because there's an import flow to migrate existing configurations.

Chumba commented:
@klaashoekstra94 Thanks so much for your work in extending Onkyo support, this is very exciting stuff. Could I very humbly ask for 1 small (I believe) addition? My AVR supports the Whole House listening mode which I believe is not included in your mappings list, is this something that you could add? I see the option in Onkyo-EISCP and I've tested it successfully which gives me hope this may be possible in your implementation. 

Thank you again

klaashoekstra94 committed:
Add additional listening modes

klaashoekstra94 commented:
@Chumba I added the whole house listening mode. Initially I had not implemented it since it does not seem to be working on my AVR, but if it works on yours we should add it. Now let's hope this PR gets merged soon!

klaashoekstra94 commented:
Hey @frenck, I was wondering, when can I expect some progress on this PR? Thanks in advance!

raman325 commented:
I don't think the PR description matches the actual PR anymore. I suggest instead of trying to pile all of these changes into one PR, you split out the config flow addition, the library switch, and the addition of sound mode into 3 separate PRs. I can almost guarantee you will get a lot faster turnaround on those. So much is changing in this PR that it's hard to keep track of everything.

klaashoekstra94 commented:
@raman325 I agree that the PR has gotten quite big, but @frenck already made some initial review comments that I resolved. Also I would greatly appreciate it if this could stay a single PR, instead of having to split everything apart again, given the time I already put into this. Also, by keeping this in a single PR, the breaking changes are minimal. Let's see what happens!

klaashoekstra94 committed:
Fix options config flow test and translation

klaashoekstra94 commented:
Hey @frenck, I'd still like to get this into Home Assistant :) Can you maybe review this, or do you need anything else from me?

klaashoekstra94 committed:
"Fix incorrect rebase of mypy.ini"

klaashoekstra94 commented:
@frenck I'm still trying to keep this PR up to date, would be great if it can be reviewed. Also I see @MartinHjelmare added this to Needs review, does that mean it will be reviewed?

frenck commented:
Be patient, thanks 👍

klaashoekstra94 committed:
Update number RestoreEntity to RestoreNumber

klaashoekstra94 commented:
Got this back up to date with the dev branch. I also changed the RestoreEntity in the number platform to a RestoreNumber, added an integration type to the manifest file, fixed an async timeout deprecation warning, updated the integrations.json and fixed a missing import.

klaashoekstra94 committed:
Fixed mypy and isort issues

BenPru reviewed (approved):
lgtm

BenPru committed:
Use Platform consts

klaashoekstra94 committed:
Fix ruff issue

klaashoekstra94 committed:
Refactor max volume number and fix warning in init

epenet commented on homeassistant/components/onkyo/translations/en.json:
```diff
@@ -0,0 +1,81 @@
+{
```
This file is no longer required - please remove it.

epenet commented on homeassistant/components/onkyo/number.py:
```diff
@@ -0,0 +1,114 @@
+"""Platform for sensor integration."""
+from __future__ import annotations
```
I find it very confusing that you are adding a new platform as part of config-flow implementation.
I think it should have been kept for a follow-up PR.

epenet commented on homeassistant/components/onkyo/manifest.json:
```diff
@@ -1,9 +1,11 @@
+  "iot_class": "local_push",
   "loggers": ["eiscp"],
-  "requirements": ["onkyo-eiscp==1.2.7"]
+  "requirements": ["pyeiscp==0.0.7"]
```
Changing the library could be done as a preliminary PR, no?

epenet commented on homeassistant/components/onkyo/config_flow.py:
```diff
@@ -0,0 +1,271 @@
+        ]
+
+
+class OptionsFlowHandler(config_entries.OptionsFlow):
+    """Handle options for the Onkyo component."""
+
+    def __init__(self, config_entry: config_entries.ConfigEntry) -> None:
+        """Initialize options flow."""
+        self.config_entry: config_entries.ConfigEntry = config_entry
+        self._options: dict[str, str | dict[str, Any]] = {}
```
```suggestion
class OptionsFlowHandler(config_entries.OptionsFlowWithConfigEntry):
    """Handle options for the Onkyo component."""

```

epenet commented on .strict-typing:
```diff
@@ -223,6 +223,7 @@ homeassistant.components.number.*
 homeassistant.components.nut.*
 homeassistant.components.oncue.*
 homeassistant.components.onewire.*
+homeassistant.components.onkyo.*
```
It's very confusing that strict-typing is implemented at the same time as a config-flow PR.
Can that not be a separate/preliminary PR?

epenet commented on tests/components/onkyo/test_config_flow.py:
```diff
@@ -0,0 +1,287 @@
+}
+
+
+def get_mock_connection(host, name, identifier):
```
It is recommended that config_flow tests do not hit the integration internals.
You should look at adding `mock_setup_entry` in a local `conftest.py` file.
```suggestion
pytestmark = pytest.mark.usefixtures("mock_setup_entry")


def get_mock_connection(host, name, identifier):
```

epenet reviewed (changes requested):
A few things I have spotted.
Mainly my recommendation is that this PR is bigger than it needs to be.
Some of the changes could be done in a preliminary or follow-up PR.
See https://developers.home-assistant.io/docs/review-process#creating-the-perfect-pr

klaashoekstra94 commented:
@epenet Sorry for the late response, work is busy at the moment so I didn't have the time yet to look into your review. I actually wanted to do the switch to the other library as a preliminary PR, but it was rejected because I was not allowed to add functionality to a component that didn't have a config flow yet, is that not the case anymore? Because that's why I also added the config flow in here, as it would be a lot of work to first make a config flow and then change the library to the async one, meaning I would have to do a big part of the config flow again.

foxy82 commented:
The onkyo failures caused by the last Home Assistant core release (https://github.com/home-assistant/core/issues/47918) reminded me of this and that the onkyo integration really needs a refresh.

I wonder if we can actually get this over the line - I've been using a version of this since 2021 when it first appeared and it is much more responsive than the official Home Assistant one. Or @klaashoekstra94 maybe we just cut our losses and move it to HACs?

klaashoekstra94 commented:
@foxy82 I would prefer having this directly in HA instead of using HACS, but I'm afraid that would mean that I have to split this PR into multiple smaller ones. Maybe I can use PR #48640 as a starting point but make sure that the YAML platform schema doesn't change at first. That way I could at least get the async library in first. Still that would mean a major overhaul to add the config flow in a second PR, making that one possibly almost just as big as this one. So I would still prefer if this one could get merged.

foxy82 commented:
I started to look at adding the config flow to the current Home Assistant version.

It needs tests and the strings aren't working properly but it's an initial version that works. I based it hugely on this pr in the hope that it it is implemented rebasing this from it shouldn't be too hard.

I'll try to push it later.

foxy82 commented:
Here is the start of the work to add config_flow to the existing component in case we can't get this PR merged.

https://github.com/foxy82/home-assistant/pull/106

shbatm commented:
Let me know if you need any help testing this.

taw123 commented:
I too would be interested in assisting in any way helpful here as I have a TX-NR636 I have for nearly 4 years used with the ancient/standard Onkyo integration. Before that/and overlapping with my HA work I did a bit of code tweaking using ToddGreenfeild's HomeBridge implementation. 

Unfortunately I'm not clear non what's required to integrate a pre-release component in HomeAssistant. Given the comments about migration above, I didn't mind recreating the integration/configuration and entities if still required to use given migration issues.

Thanks for any pointers or guidance. While not totally clueless as a former Engineer but not very familiar with the internals of HomeAssistant though willing to learn/help.

--T

foxy82 commented on homeassistant/components/onkyo/config_flow.py:
```diff
@@ -0,0 +1,271 @@
+        ]
+
+
+class OptionsFlowHandler(config_entries.OptionsFlow):
```
From what I can see having these at the level forces the user to use the same input sources across all of their receivers.

I don't think it allows the following to happen: 

Receiver 1 - I want to import Game, TV and rename Game to "PlayStation"
Receiver 2 - I want to import Game, CD, TV, Aux and rename Game to "XBox"

dawiinci commented:
> Unfortunately I'm not clear non what's required to integrate a pre-release component in HomeAssistant. Given the comments about migration above, I didn't mind recreating the integration/configuration and entities if still required to use given migration issues.


I would like to know as well, since it seems like this is going on for a long time. Also it would be nice to test if the issues are actually gone.

foxy82 commented:
> > Unfortunately I'm not clear non what's required to integrate a pre-release component in HomeAssistant. Given the comments about migration above, I didn't mind recreating the integration/configuration and entities if still required to use given migration issues.
> 
> I would like to know as well, since it seems like this is going on for a long time. Also it would be nice to test if the issues are actually gone.
> 
> Is it even possible to just alter the any part of HA manually? I guess this has to be done after every update of HA which would be fine to me if it works better like that.

@taw123 and @dawiinci I've added it as a custom HACs repo.

If you setup HACS then go HACS -> 3 Dots -> Custom Repositories.
In repositories enter "https://github.com/foxy82/hacs-onkyo-repository", Category:  "Integration" and click "Add".
Then search for "Onkyo" in HACS and install  "Onkyo/Pioneer async component"
Note I've intentionally changed the platform name from `onkyo` to `onkyo_async` to avoid clashing with the official component.  You will need to convert your existing `onkyo` yaml config to one that is compatible with `onkyo_async` there are details in the README.md in the repository. Any issues getting setup raise it in my repo above.

dawiinci commented:
Seems to work, thanks. However, if the power is switched off, the receiver still says it is on. Is there a way to update the state manually?

Switching it off after power is gone doesn't update.

foxy82 commented:
> Seems to work, thanks. However, if the power is switched off, the receiver still says it is on. Is there a way to update the state manually?
> 
> Switching it off after power is gone doesn't update.
> 
> It doesn't even update after a restart which the original does.
> 
> Unfortunately, after a restart the device is missing completely, if it is offline.
> 
> Reverting back to original fixes this behavior.

Raise these here: https://github.com/foxy82/hacs-onkyo-repository/issues - so the PR stays on track.

Daneish commented:
Hi there, Klaas thank you so much for your work implementing this modernised Onkyo integration, I’m really looking forward to it being merged.

To assist with the review process, I tried to look over the code, but not being a developer, I fear I would do more harm than good. But, for about 6 months I had been testing this code as a manually created custom component and it was working great.

Then around May of this year presumably after changes to Core, the integration would no longer load with the error:
<details>
<summary>AttributeError: 'ConfigEntries' object has no attribute 'async_setup_platforms'</summary>

2023-09-14 22:35:22.595 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry Living Room Receiver for onkyo
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/config_entries.py", line 387, in async_setup
result = await component.async_setup_entry(hass, self)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/onkyo/__init__.py", line 28, in async_setup_entry
hass.config_entries.async_setup_platforms(entry, PLATFORMS)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'ConfigEntries' object has no attribute 'async_setup_platforms'
</details>

So it looks like it may need some changes to get going again, but (from an non-dev’s perspective) it looks really solid.

I have also recorded some observations I had made when it was working, that I would be happy to share, if they would be of any value.

Franck I just wanted to flick you a quick message to thank you very much for your previous reviews on this PR. In addition to that I really appreciate all of your work to Home Assistant such a great platform! 

It looks like Klaus has been hoping to get another review from you. When you get back from your vacation and are caught up with the chaos of returning, would you please consider casting your expertise over this PR?

I really want this to comes across as a polite and genuine request. I know there are quite a few users that would really benefit from an update to the Onkyo integration and I didn’t want this PR to get closed off by the bot.

Thanks for reading and your consideration!

With respect and kindness,
Daneish

bdraco reviewed (changes requested):
https://developers.home-assistant.io/docs/creating_component_code_review#5-make-your-pull-request-as-small-as-possible

While its already been said in https://github.com/home-assistant/core/pull/59518#issuecomment-1072916645  Please break this up into smaller chunks. It should be at minimum 2 PRs (3 if the new library is mostly compatible with the old one).  This PR won't be able to move forward with the size and scope of the review required as its simply too large to review effectively.

1. Add the config flow
2. Switch out the library
3. Switch to local push

frenck commented:
Because there hasn't been any activity on this PR for quite some time now, I've decided to close it for being stale.

Feel free to re-open this PR when you are ready to pick up work on it again 👍

../Frenck
